### PR TITLE
Fix race condition in omp parallel for

### LINKED ISSUE #103: [ep1 cuda eemumu] Race condition in OMP parallel for
Due to implicit default sharing in `omp parallel for`, all threads are writing chaotically into `sigmakin_itry, sigmakin_goodhel`.

https://github.com/madgraph5/madgraph4gpu/blob/08830a88a9d0a302926c99a9807a81858a5e2bd3/epoch1/cuda/ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum/CPPProcess.cc#L699-L707

### PR DESCRIPTION
**Pull request by hageboeck**
- Fix #103, race condition in parallel for
- Fix missing include
- Fix compilation with clang.

**Note**:
Although this fixes a race condition, we can potentially still get different results in dependence of number of threads, namely if analysing events 0-9 and computing good helicities will give a different result than analysing events 100-109, for example.
I think that the physics should be robust against that, but I'm not 100% sure.

That's probably not a problem, though, since when starting 2 madgraph processes with different random numbers, you are in the same situation: Helicities get computed from different events.

### CODE REVIEW

**Review by lfield — CHANGES_REQUESTED** [review]
Doesn't build.
```
CPPProcess.cc: In function ‘void Proc::sigmaKin(const fptype*, mgOnGpu::fptype*, int)’:
CPPProcess.cc:711:5: error: ‘nevt’ not specified in enclosing ‘parallel’
  711 |     for (int ievt = 0; ievt < nevt; ++ievt)
      |     ^~~
CPPProcess.cc:710:9: error: enclosing ‘parallel’
  710 | #pragma omp parallel for default(none) shared(allmomenta, allMEs) firstprivate(sigmakin_itry, sigmakin_goodhel)
```

**Comment by hageboeck** [comment]
Funny. Which OMP/compiler version? Github (ubuntu), macos and centos8 agreed that it's OK.

**Review by lfield — APPROVED** [review]

**Comment by lfield** [comment]
Ubuntu 20.04 LTS  gcc 9.3.0. Now builds.

File: epoch1/cuda/ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum/check.cc [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #include <algorithm>
+#include <array>
```
Comment (valassi) [inline epoch1/cuda/ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum/check.cc:2]: Why was this needed actually?

File: epoch1/cuda/ee_mumu/SubProcesses/Makefile [hunk 2]
```diff
@@ -3,7 +3,8 @@ TOOLSDIR = ../../../../../tools
 INCFLAGS = -I. -I../../src -I$(TOOLSDIR)
 MODELLIB = model_sm
 OPTFLAGS = -O3
-CXXFLAGS = $(OPTFLAGS) -std=c++11 $(INCFLAGS) $(USE_NVTX) -Wall -Wshadow -Wextra -fopenmp $(MGONGPU_CONFIG)
+OMPFLAGS?= -fopenmp
```
Comment (valassi) [inline epoch1/cuda/ee_mumu/SubProcesses/Makefile:6]: Why do you want to make this optional?

**Review by valassi — COMMENTED** [review]
Hi Stephan, thanks a lot, looks good. I just have two minor questions.

I think it's a good idea to have default none and be more in control.

About the helicities, I am a bit puzzled what this does. But I would not bother too much on teh specific point: with the vectorization work discussed yesterday, the helicity loop is done in a completely different way, so I think that the issue you mentioned will no longer exist.

Approved, but maybe answer on the two points please? thanks Andrea

File: epoch1/cuda/ee_mumu/SubProcesses/Makefile (hunk 2 above)
Comment (hageboeck) [reply epoch1/cuda/ee_mumu/SubProcesses/Makefile:6]: It's a workaround for clang. Its compiler driver doesn't understand `-fopenmp`, but the C++ backend does.

The details are in the commit message of b419dffdb3428ed0ca72d44767ab25f987ae7cc6

File: epoch1/cuda/ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum/check.cc (hunk 1 above)
Comment (hageboeck) [reply epoch1/cuda/ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum/check.cc:2]: I don't know who secretly includes `array` in the other setups, but it's certainly not included in `check.cc`.
```
P1_Sigma_sm_epem_mupmum (epoch2_tests)$ make
Makefile:42: No cuda found. Export CUDA_HOME to compile with cuda
clang++  -O3 -std=c++11 -I. -I../../src -I../../../../../tools  -Wall -Wshadow -Wextra -Xclang -fopenmp -lomp -DMGONGPU_COMMONRAND_ONHOST  -c check.cc -o check.o
check.cc:602:27: error: implicit instantiation of undefined template 'std::__1::array<char, 128>'
    std::array<char, 128> nprocbuf;
```

Comment (hageboeck) [reply epoch1/cuda/ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum/check.cc:2]: I would actually bet on a bug in the standard library, i.e. some header also includes secretly includes array.

File: epoch1/cuda/ee_mumu/SubProcesses/Makefile (hunk 2 above)
Comment (valassi) [reply epoch1/cuda/ee_mumu/SubProcesses/Makefile:6]: Thanks! Sorry I should have said before, I had seen some similar issues in clang, this is how I fixed it, does this help?
https://github.com/valassi/madgraph4gpu/commit/97ac750deb148ba89125ed128bbc50ae62074096

Comment (valassi) [reply epoch1/cuda/ee_mumu/SubProcesses/Makefile:6]: Essentially -lomp instead of -lgomp, for runTest linking

**Comment by lfield** [comment]
I ran this code through the Intel Advisor and it suggested to use the following for the for loop in sigmaKin.
```
#pragma omp teams distribute parallel for simd map(to: ) map(from: )
```

File: epoch1/cuda/ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum/check.cc (hunk 1 above)
Comment (valassi) [reply epoch1/cuda/ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum/check.cc:2]: Thanks. Funny I also tried clang yesterday and I did not see that (I saw a few warnings instead).

**Comment by valassi** [comment]
Thanks Laurence. I am not sure simd would do much without more significant code changes, but you never know. In case you want to play a bit more with Intel Advisor, you can try my klas branch from #72, there I have a factor 6 speedup (4 from AVX2, plus 1.5 from maybe pipelining), it would be useful to udnerstand better with a profiler. It's usable, but still rapidly changing WIP... Wait a few days otherwise.

File: epoch1/cuda/ee_mumu/SubProcesses/Makefile (hunk 2 above)
Comment (hageboeck) [reply epoch1/cuda/ee_mumu/SubProcesses/Makefile:6]: This seem to be orthogonal. My clang driver really doesn't swallow `-fopenmp`. It says that there's no such parameter.

Comment (valassi) [reply epoch1/cuda/ee_mumu/SubProcesses/Makefile:6]: Uh strange. I used the SFT installed clang. This is what I get in my clang branch with just those two changes (plus your merges I think)

```
[avalassi@itscrd03 bash] ~/GPU2020/madgraph4gpu/epoch1/cuda/ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum> make
/usr/local/cuda-11.1/bin/nvcc  -O3 -lineinfo -std=c++14 -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.1/include/ -DUSE_NVTX -arch=compute_70 -use_fast_math  -c gcheck.cu -o gcheck.o
/usr/local/cuda-11.1/bin/nvcc  -O3 -lineinfo -std=c++14 -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.1/include/ -DUSE_NVTX -arch=compute_70 -use_fast_math  -c gCPPProcess.cu -o gCPPProcess.o
/usr/local/cuda-11.1/bin/nvcc gcheck.o -o gcheck.exe gCPPProcess.o -arch=compute_70 -L../../lib -lmodel_sm -L/usr/local/cuda-11.1/lib64/ -lcuda -lcurand
/cvmfs/sft.cern.ch/lcg/releases/clang/10.0.0-62e61/x86_64-centos7/bin/clang++  -O3 -std=c++11 -I. -I../../src -I../../../../../tools -DUSE_NVTX -Wall -Wshadow -Wextra -fopenmp  -I/usr/local/cuda-11.1/include/ -c check.cc -o check.o
/cvmfs/sft.cern.ch/lcg/releases/clang/10.0.0-62e61/x86_64-centos7/bin/clang++  -O3 -std=c++11 -I. -I../../src -I../../../../../tools -DUSE_NVTX -Wall -Wshadow -Wextra -fopenmp  -I/usr/local/cuda-11.1/include/ -c CPPProcess.cc -o CPPProcess.o
In file included from CPPProcess.cc:444:
./CPPProcess.h:86:9: warning: private field 'm_numiterations' is not used [-Wunused-private-field]
    int m_numiterations;
        ^
1 warning generated.
/cvmfs/sft.cern.ch/lcg/releases/clang/10.0.0-62e61/x86_64-centos7/bin/clang++ check.o -o check.exe CPPProcess.o  -O3 -std=c++11 -I. -I../../src -I../../../../../tools -DUSE_NVTX -Wall -Wshadow -Wextra -fopenmp  -ldl -pthread -L../../lib -lmodel_sm -L/usr/local/cuda-11.1/lib64/ -lcuda -lcurand
/cvmfs/sft.cern.ch/lcg/releases/clang/10.0.0-62e61/x86_64-centos7/bin/clang++  -O3 -std=c++11 -I. -I../../src -I../../../../../tools -I../../../../../tools/googletest/googletest/include/ -DUSE_NVTX -Wall -Wshadow -Wextra -fopenmp  -I/usr/local/cuda-11.1/include/ -c runTest.cc -o runTest.o
ln -sf runTest.cc runTest_tmp.cu
/usr/local/cuda-11.1/bin/nvcc -o runTest.exe CPPProcess.o runTest.o gCPPProcess.o runTest_tmp.cu  -O3 -lineinfo -std=c++14 -I. -I../../src -I../../../../../tools -I../../../../../tools/googletest/googletest/include/ -I/usr/local/cuda-11.1/include/ -DUSE_NVTX -arch=compute_70 -use_fast_math  -ldl -L../../lib -lmodel_sm -L../../../../../tools/googletest/build/lib// -lgtest -lgtest_main -L/cvmfs/sft.cern.ch/lcg/releases/clang/10.0.0-62e61/x86_64-centos7/lib -L/usr/local/cuda-11.1/lib64/ -lcuda -lcurand -lcuda -lomp
unlink runTest_tmp.cu
```

This is from `/cvmfs/sft.cern.ch/lcg/releases/clang/10.0.0/x86_64-centos7/setup.sh`

**Pull request merged**